### ISSUE TITLE
Display error with NetEase Cloud Music

### ISSUE THREAD

**Issue by GirlKiller512** [Creepercdn/now_playing#8]
The music title is displayed incorrectly if it includes 1 or more "-" with no spaces next to "-".
Display text setting: `%title - %artist`
Examples:
![image](https://github.com/rsp4jack/now_playing/assets/20424975/7af99c1f-a83b-401a-b418-80a9d5642e53)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/b8c21f46-89af-4268-b06e-d478e7783ff6)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/ec7a5839-e318-455b-bfd7-c31f77cce150)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/d2964c16-91dc-449c-83e7-34e331f94454)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/fcb28172-1cf0-499a-b70e-5398a460156d)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/a70503e4-f23e-480c-968f-8516a9aac478)

If the music title includes "-" with 1 space on both sides of "-" or only the music artist includes "-" then it is displayed with no errors.
Display text setting: `%title - %artist`
Examples:
![image](https://github.com/rsp4jack/now_playing/assets/20424975/08102512-c431-44f2-ad9d-6434b9bf87e0)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/a5c6b5f5-adc9-4273-bb6f-72d9e7022595)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/a95a2e8a-5505-4bcb-a46a-2e0efc2759ff)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/152efd43-2e2b-42be-834d-f4cd956d3932)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/a6e646d1-25f5-431a-b2fc-dd1b4a85fd80)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/64ddd536-12e8-44d1-903a-e26d99c1625d)

**Comment by rsp4jack** [comment]
Sounds interesting. Could you give some examples with NCM window title presented?

**Comment by GirlKiller512** [comment]
These are the window titles of the music above.
![image](https://github.com/rsp4jack/now_playing/assets/20424975/12bac3de-bbff-41e8-9c95-2985049cdb82)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/a1f73030-bf7a-4271-9ee6-f5f280a14656)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/91e35e87-49a0-46d6-8f3e-3d508cb7003b)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/bed69341-ce2c-4c43-84a1-4a17b7449827)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/1e48178b-3a36-4039-81d2-315716cbbc72)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/c9182120-9940-498b-91f8-793468b32abb)

In other words, the music with a title including "-" may cause display errors.

**Comment by GirlKiller512** [comment]
If I change the display text setting to `%artist - %title`, the text becomes mess.
![image](https://github.com/rsp4jack/now_playing/assets/20424975/6e8ce7ba-b4f7-4f3c-aeed-19d6f49e3711)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/1a4a90e2-9a6b-490a-946b-c4daa8cdf425)

**Comment by rsp4jack** [comment]
I do not think there is a simple way to fix this.

Using SMTC should work (not implemented so far 😭), but it needs extra setup (`winrt` requires Python 3.7 or above, but OBS supports only Python 3.6)

**Comment by GirlKiller512** [comment]
Let's leave the issue open until a solution is found.

**Comment by rsp4jack** [comment]
Edit: I found the latest OBS supports Python 3.6 to 3.10. SMTC support is coming soon 🎉

**Comment by GirlKiller512** [comment]
Yeah, I also found that OBS 30.1.0 supports Python 3.6-3.10 from its official website last night.
![image](https://github.com/rsp4jack/now_playing/assets/20424975/b05b9e9f-da5a-4f8a-8054-7a746b7c4b07)

**Comment by rsp4jack** [comment]
I think it works now. Just use SMTC capture only and everything will be fine 😺

This also breaks compatibility for older Windows versions (only Windows 10 and 11 support SMTC now)

**Comment by rsp4jack** [comment]
BTW, SMTC seems to be slow and sometimes it will time out. I think that is normal unless every SMTC query operation is timed out.

**Comment by GirlKiller512** [comment]
`winrt-Windows.Foundation` and `winrt-Windows.Media.Control` seem to require Python 3.9+ so I only tested the script with 3.9 and 3.10. All dependencies have been installed on both versions.

On Python 3.10, the script can be loaded but it keeps showing errors every 1 second in the script log.
Settings: 
![image](https://github.com/rsp4jack/now_playing/assets/20424975/2e52415a-606b-4da3-9f07-639e07240a0e)
```
[now_playing.py] [18:14:35] [MainThread/DEBUG]: [proactor_events]: Using proactor: IocpProactor
[now_playing.py] [18:14:35] [MainThread/DEBUG]: [now_playing]: script_defaults(<Swig Object of type 'obs_data_t *' at 0x0000029EBD878F00>)
[now_playing.py] [18:14:35] [MainThread/DEBUG]: [now_playing]: script_description()
[now_playing.py] [18:14:35] [MainThread/INFO]: [now_playing]: script_load()
[now_playing.py] [18:14:35] [MainThread/DEBUG]: [now_playing]: script_update(<Swig Object of type 'obs_data_t *' at 0x0000029EBD878F00>)
[now_playing.py] [18:14:37] [MainThread/DEBUG]: [now_playing]: script_properties()
[now_playing.py] [18:14:37] [MainThread/INFO]: [now_playing]: locale: zh-CN
[now_playing.py] [18:14:39] [MainThread/DEBUG]: [now_playing]: script_update(<Swig Object of type 'obs_data_t *' at 0x0000029EBEDC98F0>)
[now_playing.py] [18:14:39] [MainThread/DEBUG]: [now_playing]: Timer started
[Unknown Script] [18:14:40] [nowplaying_eventloop/DEBUG]: [now_playing]: smtcCaptureAsync takes 0.005567073822021484s
[Unknown Script] [18:14:40] [nowplaying_eventloop/DEBUG]: [now_playing]: doUpdate: []
[Unknown Script] [18:14:40] [nowplaying_eventloop/ERROR]: [now_playing]: onUpdate error
[Unknown Script] Traceback (most recent call last):
[Unknown Script]   File "D:\Documents/Downloads\now_playing.py", line 308, in doUpdate
[Unknown Script]     update_song(data[0])
[Unknown Script] IndexError: list index out of range
[Unknown Script] [18:14:41] [nowplaying_eventloop/DEBUG]: [now_playing]: smtcCaptureAsync takes 0.002000570297241211s
[Unknown Script] [18:14:41] [nowplaying_eventloop/DEBUG]: [now_playing]: doUpdate: []
[Unknown Script] [18:14:41] [nowplaying_eventloop/ERROR]: [now_playing]: onUpdate error
[Unknown Script] Traceback (most recent call last):
[Unknown Script]   File "D:\Documents/Downloads\now_playing.py", line 308, in doUpdate
[Unknown Script]     update_song(data[0])
[Unknown Script] IndexError: list index out of range
[Unknown Script] [18:14:42] [nowplaying_eventloop/DEBUG]: [now_playing]: smtcCaptureAsync takes 0.002012968063354492s
[Unknown Script] [18:14:42] [nowplaying_eventloop/DEBUG]: [now_playing]: doUpdate: []
[Unknown Script] [18:14:42] [nowplaying_eventloop/ERROR]: [now_playing]: onUpdate error
[Unknown Script] Traceback (most recent call last):
[Unknown Script]   File "D:\Documents/Downloads\now_playing.py", line 308, in doUpdate
[Unknown Script]     update_song(data[0])
[Unknown Script] IndexError: list index out of range
[now_playing.py] [18:14:43] [MainThread/DEBUG]: [now_playing]: script_update(<Swig Object of type 'obs_data_t *' at 0x0000029EBED07900>)
[now_playing.py] [18:14:43] [MainThread/DEBUG]: [now_playing]: Timer killed
```

On Python 3.9, the script can not be loaded.
```
[now_playing.py] [18:06:23] [MainThread/DEBUG]: [proactor_events]: Using proactor: IocpProactor
[now_playing.py] Traceback (most recent call last):
[now_playing.py]   File "D:\Documents/Downloads\now_playing.py", line 273, in <module>
[now_playing.py]     loopthread: threading.Thread | None = None
[now_playing.py] TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
```

NetEase Cloud Music is running and playing music in the background during testing.

**Comment by GirlKiller512** [comment]
Forget to provide my environment:
Windows 11 Pro 23H2 (10.0.22631.3296)
OBS 30.1.0
NetEase Cloud Music 3.0.0

**Comment by rsp4jack** [comment]
I have no idea about fixing this issue rather than SMTC.

NCM does not have SMTC support. BetterNCM got a plugin called InfLink that adds SMTC support to NCM.
But it seems not to work for NCM 3.0.0. Downgrading NCM to 2.x.x will work for InfLink.

You may use SMTC compatible players.

**Comment by rsp4jack** [comment]
`IndexError: list index out of range` error means there is no playing data that the script gets. And the script does not support Python 3.9 at all.

**Comment by GirlKiller512** [comment]
I downgraded NetEase Cloud Music to 2.10.13 and installed BetterNCM and InfLink.
Although the script has a slight delay, it displays the text correctly.
![image](https://github.com/rsp4jack/now_playing/assets/20424975/51c4e0f5-abf1-417e-b70a-f8fffee75aa2)
![image](https://github.com/rsp4jack/now_playing/assets/20424975/d6c7db2f-12e2-4b19-b304-efaef8d491c3)
Thanks for your work!

By the way, how to make the script log dialog not pop up on OBS startup with the script enabled?
I don't want to see a dialog pop up every time I launch OBS.
And, if I close NCM, the dialog will also pop up, showing `IndexError: list index out of range`.

**Comment by rsp4jack** [comment]
I replaced "debug logging" to "log level" so it is possible now to silence logging of the plugin at all, and the log dialog will not popup as there will be no log if logging level is set to "SILENT".
The IndexError will be also handled gracefully now.

**Comment by rsp4jack** [comment]
BTW, you may use a 3rd-party NCM client like LyricEase which supports SMTC and has a nice UI.

**Comment by GirlKiller512** [comment]
I'll temporarily use NetEase Cloud Music + BetterNCM combination for a period of time. The script now works great.
Thanks again for your work.
